nntrainer/layers/embedding.cpp

```diff
@@ -133,10 +135,41 @@ void EmbeddingLayer::copy(std::shared_ptr<Layer> l) {
 }
 
 void EmbeddingLayer::calcDerivative() {
-  // NYI
+  // Uncomment this after fixing issues backwarding of first layer. (Issues
+  // #1017)
+  // throw exception::not_supported(
+  //   "calcDerivative for Embedding layer is not supported");
+  return; // intended
 }
+
 void EmbeddingLayer::calcGradient() {
-  // NYI
+  Tensor &djdw =
+    weightAt(static_cast<int>(EmbeddingParams::weight)).getGradientRef();
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  djdw.setZero();
+
+  // TODO:
+  // This is to calculate gradient with current implementation of optimizer.
+  // In order to accelerate, we need to better way like using index to weight.
+
+  for (unsigned int b = 0; b < input_.batch(); ++b) {
+    float *in_data = input_.getAddress(b * input_.getDim().getFeatureLen());
+
+    for (unsigned int i = 0; i < in_length; ++i) {
+      if (in_data[i] - 1 < 0)
+        continue;
```

Comment:
Cause if padding is exist then it will be 0 until the end of in_data, so I think break will be better than continue.

Reply:
when using break, you also need to think about masking the inputs for future.